### DISCUSSION TITLE
Copy custom theme to image via a Dockerfile

### DISCUSSION THREAD

**Discussion by Calybyte** [keycloak/keycloak#23941] [Q&A] [answered]
Hi guys,

I'm trying to deploy a custom theme to Keycloak, and I'm having some issues with the Dockerfile.
In Azure Devops I created a repository with the following contents:

```
\themes\my_customtheme\common\...
\themes\my_customtheme\email\...
\themes\my_customtheme\login\...
\themes\my_customtheme\welcome\...
\azure_pipeline.yml
\Dockerfile
\readme.md
```

The 'my_customtheme' folder and its subfolders and files are based on one of the standard Keycloak themes. To doublecheck the theme is OK, I followed the tutorial at https://medium.com/@oruchan.asar/customizing-the-login-page-of-keycloak-docker-image-eefac226fd7a
This worked; I could select the theme in the dropdown menu of the realm theme settings.

I created the following Dockerfile to deploy Keycloak and copy the custom theme to the theme folder in the Docker image:

```
FROM quay.io/keycloak/keycloak:latest as builder

ENV KC_HEALTH_ENABLED=true
ENV KC_METRICS_ENABLED=true

WORKDIR /opt/keycloak
RUN /opt/keycloak/bin/kc.sh build

FROM alpine/curl:8.1.2 as digicert
WORKDIR /app
RUN curl https://dl.cacerts.digicert.com/DigiCertGlobalRootCA.crt.pem -o DigiCertGlobalRootCA.crt.pem

FROM quay.io/keycloak/keycloak:latest
COPY --from=builder /opt/keycloak/ /opt/keycloak/
COPY --from=digicert /app/DigiCertGlobalRootCA.crt.pem /opt/keycloak/.postgresql/root.crt

COPY themes/my_customtheme/. /opt/jboss/keycloak/themes/my_customtheme

ENTRYPOINT ["/opt/keycloak/bin/kc.sh"]
```

Devops Pipelines can build an image from this Dockerfile and deploy to Azure App Services. Keycloak runs fine, but the custom theme doesn't show up in the dropdown menu.
I'm kind of new to Dockerfiles, so it's probably something obvious I'm missing. Could you please take a look at it? Thanks in advance.

Cheers,

Paul

**Comment by kwame-mintah** [comment, marked as answer]
@Calybyte, Your Docker file is fine, it just the COPY command needs to be changed. 

The tutorial you linked in your post was for the Wildfly distribution of Keycloak, which is no longer supported as of Keycloak v20. Because you're building your docker image with the latest, you would be using the Quarkus distribution and the latest at the time of writing would be v22 as per tags:

https://quay.io/repository/keycloak/keycloak?tab=tags

```
FROM quay.io/keycloak/keycloak:latest
```

I would suggest pinning to a specific version of Keycloak or else every time your Azure Pipeline builds, it will pull the _latest_, that Keycloak owners have pointed it to, so would change the version to e.g.
```
FROM quay.io/keycloak/keycloak:22.0.0
```

So, the theme copy, it would look like this:
```
COPY themes/my_customtheme/. /opt/keycloak/themes/my_customtheme
```

Notice that the path has changed between Wildfly (JBoss) and Quarkus.

**Reply by Calybyte** [reply]
Thank you! It was indeed the wrong destination path.
